provider/zone_id_filter.go

```diff
@@ -0,0 +1,45 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import "strings"
+
+// ZoneIDFilter holds a list of zone ids to filter by
+type ZoneIDFilter struct {
+	zoneIDs []string
+}
+
+// NewZoneIDFilter returns a new ZoneIDFilter given a list of zone ids
+func NewZoneIDFilter(zoneIDs []string) ZoneIDFilter {
+	return ZoneIDFilter{zoneIDs}
+}
+
+// Match checks whether a zone matches one of the provided zone ids
+func (f ZoneIDFilter) Match(zoneID string) bool {
+	// An empty filter includes all zones.
+	if len(f.zoneIDs) == 0 {
+		return true
+	}
+
+	for _, id := range f.zoneIDs {
+		if strings.HasSuffix(zoneID, id) {
```

Comment:
@vboginskey Why is this added as a `HasSuffix` this is currently causing us issues. We had zones called `zone` and `public-zone` which means that the first one also matches the last one. One would think that ID would be unique and hence an equal operation would be better. Are there some use-cases for this, or is it just a 'bug'?

Reply:
I implemented it this way to avoid confusing users. For Route 53, zone IDs have the form `/hostedzone/Z119WBBTVP5WFX`. However, in the AWS console they are reported as `Z119WBBTVP5WFX`. If someone supplied the latter as the zone ID and I used an exact match, it would fail to match.